File: 226. Invert Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // DFS: 40.54%, ok
    TreeNode* invertTree(TreeNode* root) {
        if(!root) return root;
        dfs(root);
        return root;
    }

    void dfs(TreeNode* root){
        if(!root->left && !root->right) return;
        TreeNode* temp = root->right;
        root->right = root->left;
        root->left = temp;
        if(root->left) dfs(root->left);
        if(root->right) dfs(root->right);
    }

    // DFS without auxilary function: 40.54%, ok
    TreeNode* invertTree(TreeNode* root){
        if(!root) return root;
        root->left = invertTree(root->left);
        root->right = invertTree(root->right);
        swap(root->left, root->right);
        return root;
    }
    
    // iterative/BFS: O(Nodes), 40.54%, ok
    TreeNode* invertTree(TreeNode* root) {
        if(!root) return NULL;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            TreeNode* temp = q.front();
            q.pop();
            swap(temp->left, temp->right);
            if(temp->left) q.push(temp->left);
            if(temp->right) q.push(temp->right);
        }
        return root;
    }

};

File: 226. Invert Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # dfs: O(N) time, O(N) space
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root:
            root.left, root.right = root.right, root.left
            self.invertTree(root.left)
            self.invertTree(root.right)
        return root
    
    # bfs: O(N) time, O(N) space
    def invertTree(self, root):
        if root == None:    # corner case
            return None
        queue = [root]
        for node in queue:
            node.left, node.right = node.right, node.left
            queue += filter(None, [node.left, node.right])
        return root
                
        